content/algorithms/isomorphism/isomorphism.md

```diff
@@ -0,0 +1,244 @@
+- G and H have the same amount of nodes and edges 
+- The degree sequence for G and H are the same
+
+These are necessary conditions but don't guarantee that two graphs are isomorphic. Let's see a small example:
+
+```{code-cell}
+plt.subplot(221)
+G = nx.cycle_graph(6)
+nx.draw_circular(G)
+plt.title("G", fontweight="bold")
+plt.subplot(222)
+H = nx.union(nx.cycle_graph(3), nx.cycle_graph(3), rename=("s", "d"))
+nx.draw_circular(H, node_color="r")
+plt.title("H", fontweight="bold")
+plt.show()
+```
+
+We can use the function *faster_could_be_isomorphic()* that return True if G and H have the same degree secuence.
+
+```{code-cell}
+nx.faster_could_be_isomorphic(G, H)
+```
+
+These graphs are clearly not isomorphic but they have the same degree secuence.
+
+
+Another property we can check for is: 
+- The same number of cycles of a particular length, for example, triangles. 
+
+We can use the function *fast_could_be_isomorphic()* to check if the graphs have the same degree and triangle sequence. The triangle sequence contains the number of triangles each node is part of.
+
+```{code-cell}
+nx.fast_could_be_isomorphic(G, H)
+```
+
+This new property allows us to detect that the graphs from the previous example were not isomorphic. 
+
+We can go one step further and check: 
+- The same number of maximal cliques. 
+
+We can use the function *could_be_isomorphic()* to check if the graphs have the same degree, triangle, and clique sequence. The clique sequence contains for each node the number of the maximal clique involving that node.
+
+```{code-cell}
+nx.could_be_isomorphic(G, H)
+```
+
+Again we can detect that G and H are not isomorphic. But these conditions are not enough to say that two graphs are isomorphic. Let's look at the following example:
+
+```{code-cell}
+plt.subplot(221)
+G = nx.path_graph(5)
+G.add_edge(2, 5)
+nx.draw_circular(G, with_labels=True, node_color="g")
+plt.title("G", fontweight="bold")
+
+plt.subplot(222)
+H = nx.path_graph(5)
+H.add_edge(3, 5)
+nx.draw_circular(H, with_labels=True, node_color="c")
+plt.title("H", fontweight="bold")
+plt.show()
+```
+
+```{code-cell}
+nx.could_be_isomorphic(G, H)
+```
+
+These graphs meet all the necessary conditions but they're not isomorphic.
+
+
+## Some classes of graphs with solution in polynomial time
+- Trees
+- Planar graphs(In fact, planar graph isomorphism is O(log(n)))
+- Interval graphs
+- Permutation graphs
+- Circulant graphs
+- Bounded-parameter graphs
+    - Graphs of bounded treewidth
+    - Graphs of bounded genus
+    - Graphs of bounded degree
+    - Graphs with bounded eigenvalue multiplicity
+    - k-Contractible graphs (a generalization of bounded degree and bounded genus)
+
+
+Let's see an example, we can use the function *tree_isomorphism()* from the isomorphism module to check if two trees are isomorphic in $O(n*log(n))$. This function uses a D&C approach to match the trees once it has found the root of each tree and returns a list with the node matching. 
```

Comment:
I've never heard of `D&C approach` - we should probably more explicitly state what this means!